### PR TITLE
fix: implement getStructure method for S3 plugin

### LINKED ISSUE #8857: [Bug]: S3 / UQI:  Implement `get structure` method for S3.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

- Implement `get structure` method for S3. 
  - Add templates for command generation.
  - Update flow for `generate page`.

### Steps To Reproduce

1. Create S3 datasource.
2. Check that the structure in entity explorer is empty. 

### Environment

Production

### Version

Cloud

### PR DESCRIPTION
**Pull request by sumitsum**
## Description
- Add `getStructure` implementation for S3 to return a list of accessible buckets. 
- Add command templates for S3. 
  - List files
  - Read file
  - Create file
  - Delete file
- Remove `getDatasourceMetadata` implementation as `getStructure` method can be directly used now. 
- TBD: Some client side implementations / fixes are required before this code can be merged:
  - Currently the template is rendered correctly but the default value for `command` UQI attribute is not picked up by the client. 
     - Update: Isolating this issue out into a new issue: #9455.
  - Need to support picking up values from `ActionConfiguration` object in the `Template` object. 
  - Update client to use `getStructure` method call for S3 plugin instead of `getDatasourceMetadata`.

Fixes #8857 

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- JUnit TC
- Manual

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/s3_get_structure 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.1 **(0)** | 36.58 **(-0.01)** | 34.54 **(0)** | 55.61 **(0)**
 :green_circle: | app/client/src/pages/Editor/GeneratePage/components/GeneratePageForm/GeneratePageForm.tsx | 18.42 **(0.24)** | 10.53 **(0.06)** | 0 **(0)** | 20.59 **(0.3)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**</details>

### CODE REVIEW

File: app/server/appsmith-plugins/amazons3Plugin/src/main/java/com/external/plugins/AmazonS3Plugin.java [hunk 1]
```diff
@@ -538,14 +539,6 @@ String readFile(AmazonS3 connection, String bucketName, String path, Boolean enc
                         connection.deleteObject(bucketName, path);
                         actionResult = Map.of("status", "File deleted successfully");
                         break;
-                    case LIST_BUCKETS:
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/amazons3Plugin/src/main/java/com/external/plugins/AmazonS3Plugin.java:541]: Why are we removing this @sumitsum

**Review by nidhi-nair — APPROVED** [review]

**Comment by nidhi-nair** [comment]
Reopening because this was being worked on actively. @Irongade can you confirm whether the changes on your end are ready to be reviewed?

**Comment by sumitsum** [comment]
Thanks @nidhi-nair . This is being worked upon actively. @Irongade had added some changes for this earlier. We need to do integration testing for this now.

**Comment by Irongade** [comment]
> Reopening because this was being worked on actively. @Irongade can you confirm whether the changes on your end are ready to be reviewed?

Yes they are.

**Review by nidhi-nair — APPROVED** [review]

File: app/client/src/entities/Datasource/index.ts [hunk 2]
```diff
@@ -29,6 +29,7 @@ export interface DatasourceStructure {
 }
 
 export interface QueryTemplate {
+  actionConfiguration?: Record<string, unknown>;
```
Comment (ayushpahwa) [inline app/client/src/entities/Datasource/index.ts:32]: Use 'ActionConfig' as Type instead on this

**Review by ayushpahwa — CHANGES_REQUESTED** [review]

File: app/client/src/entities/Datasource/index.ts (hunk 2 above)
Comment (Irongade) [reply app/client/src/entities/Datasource/index.ts]: Done.

File: app/client/src/pages/Editor/GeneratePage/components/GeneratePageForm/GeneratePageForm.tsx [hunk 3]
```diff
@@ -244,7 +244,7 @@ function GeneratePageForm() {
   const {
     bucketList,
     failedFetchingBucketList,
-    fetchBucketList,
+    // fetchBucketList,
```
Comment (ayushpahwa) [inline app/client/src/pages/Editor/GeneratePage/components/GeneratePageForm/GeneratePageForm.tsx:247]: Remove this if it is not needed.

**Review by ayushpahwa — CHANGES_REQUESTED** [review]

File: app/client/src/pages/Editor/GeneratePage/components/GeneratePageForm/GeneratePageForm.tsx (hunk 3 above)
Comment (Irongade) [reply app/client/src/pages/Editor/GeneratePage/components/GeneratePageForm/GeneratePageForm.tsx]: Done.

**Review by ayushpahwa — APPROVED** [review]

**Review by nidhi-nair — DISMISSED** [review]

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**